src/Npgsql/NpgsqlReadBuffer.cs

```diff
@@ -220,7 +219,7 @@ async Task EnsureLong()
 
                             // Note that if PG cancellation fails, the exception for that is already logged internally by CancelRequest.
                             // We simply continue and throw the timeout one.
-                            if (!wasCancellationRequested && Connector.CancelRequest(requestedByUser: false))
+                            if (!HardCancellationBlock.Entered && !wasCancellationRequested && Connector.CancelRequest(requestedByUser: false))
```

Comment:
Swap two first conditions to avoid accessing the async local. It's cheaper to read the variable first.